### ISSUE TITLE
curl was removed(?), causing an RVM download problem

### ISSUE THREAD

**Issue by cvrebert** [travis-ci/travis-ci#5092]
This literally stopped working just a few minutes ago.
* [Working build](https://travis-ci.org/twbs/bootstrap/jobs/90644600)
* [Failed build log:](https://travis-ci.org/twbs/bootstrap/jobs/90644599)
```
$ rvm install 2.0.0 && rvm use 2.0.0
Searching for binary rubies, this might take some time.
Found remote file https://rvm_io.global.ssl.fastly.net/binaries/ubuntu/12.04/x86_64/ruby-2.0.0-p598.tar.bz2
Checking requirements for ubuntu.
Requirements installation successful.
ruby-2.0.0-p598 - #configure
ruby-2.0.0-p598 - #download
RVM requires 'curl'. Install 'curl' first and try again.
There was an error(200).
Checking fallback: ftp://rvm_io.global.ssl.fastly.net/binaries/ubuntu/12.04/x86_64/ruby-2.0.0-p598.tar.bz2?rvm=1.26.8
Checking fallback: http://www.mirrorservice.org/sites/rvm_io.global.ssl.fastly.net/binaries/ubuntu/12.04/x86_64/ruby-2.0.0-p598.tar.bz2?rvm=1.26.8
No fallback URL could be found, try increasing timeout with:
    echo "export rvm_max_time_flag=20" >> ~/.rvmrc
Downloading https://rvm_io.global.ssl.fastly.net/binaries/ubuntu/12.04/x86_64/ruby-2.0.0-p598.tar.bz2 failed.
Mounting remote ruby failed with status 2, trying to compile.
Checking requirements for ubuntu.
Requirements installation successful.
Installing Ruby from source to: /home/travis/.rvm/rubies/ruby-2.0.0-p598, this may take a while depending on your cpu(s)...
ruby-2.0.0-p598 - #downloading ruby-2.0.0-p598, this may take a while depending on your connection...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 10.2M  100 10.2M    0     0  46.1M      0 --:--:-- --:--:-- --:--:-- 53.7M
ruby-2.0.0-p598 - #extracting ruby-2.0.0-p598 to /home/travis/.rvm/src/ruby-2.0.0-p598....
ruby-2.0.0-p598 - #configuring..................................................
ruby-2.0.0-p598 - #post-configuration..
ruby-2.0.0-p598 - #compiling.............................................................................
ruby-2.0.0-p598 - #installing..............................
ruby-2.0.0-p598 - #making binaries executable..
RVM requires 'curl'. Install 'curl' first and try again.
There was an error while trying to resolve rubygems version for 'latest'. 
Halting the installation.
The command "rvm install 2.0.0 && rvm use 2.0.0" failed and exited with 4 during .
```
Obviously, falling back to compiling Ruby from source should be avoided if at all possible.

**Comment by cvrebert** [comment]
The error has mysteriously stopped occurring.

**Comment by cvrebert** [comment]
It's still happening: https://travis-ci.org/twbs/bootstrap/jobs/90976058
Apparently it's intermittent.

**Comment by cvrebert** [comment]
@BanzaiMan Help?

**Comment by BanzaiMan** [comment]
This is peculiar. My current guess is that the instance used in the failing builds were somehow lacking `curl` or had `$PATH` wrong.

**Comment by cvrebert** [comment]
Okay, it gets even weirder. curl is installed on the system, despite what RVM claims.
From the debug output added to https://travis-ci.org/twbs/bootstrap/jobs/91526407 :
```
$ echo $PATH
/home/travis/.nvm/versions/node/v4.2.2/bin:/home/travis/bin:/home/travis/.local/bin:/home/travis/.gimme/versions/go1.4.1.linux.amd64/bin:/usr/local/phantomjs/bin:./node_modules/.bin:/usr/local/maven-3.2.5/bin:/usr/local/clang-3.4/bin:/home/travis/.gimme/versions/go1.4.1.linux.amd64/bin:/usr/local/phantomjs/bin:./node_modules/.bin:/usr/local/maven-3.2.5/bin:/usr/local/clang-3.4/bin:/home/travis/.gimme/versions/go1.4.1.linux.amd64/bin:/home/travis/.rvm/gems/ruby-1.9.3-p551/bin:/home/travis/.rvm/gems/ruby-1.9.3-p551@global/bin:/home/travis/.rvm/rubies/ruby-1.9.3-p551/bin:/usr/local/phantomjs/bin:./node_modules/.bin:/usr/local/maven-3.2.5/bin:/usr/local/clang-3.4/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/travis/.rvm/bin:/home/travis/.rvm/bin:/home/travis/.rvm/bin
$ which curl || true
/usr/bin/curl
$ ls -l /usr/bin/curl || true
-rwxr-xr-x 1 root root 129696 Jan 14  2015 /usr/bin/curl
$ rvm install 2.2
Searching for binary rubies, this might take some time.
Found remote file https://rvm_io.global.ssl.fastly.net/binaries/ubuntu/12.04/x86_64/ruby-2.2.0.tar.bz2
Checking requirements for ubuntu.
Requirements installation successful.
ruby-2.2.0 - #configure
ruby-2.2.0 - #download
RVM requires 'curl'. Install 'curl' first and try again.
There was an error(200).
```

**Comment by cvrebert** [comment]
Still repros even after upgrading to rvm 1.26.11: https://travis-ci.org/twbs/bootstrap/jobs/91533567

**Comment by cvrebert** [comment]
Filed https://github.com/rvm/rvm/issues/3559 since the evidence points towards a possible bug in RVM.

**Comment by XhmikosR** [comment]
This still happens. If I clear the cache it works though.

**Comment by XhmikosR** [comment]
Example, failing https://travis-ci.org/mpc-hc/mpc-hc.org/builds/93137684
with cache cleared https://travis-ci.org/mpc-hc/mpc-hc.org/builds/93137684

Same build, it's just the cache is cleared.

**Comment by BanzaiMan** [comment]
@XhmikosR Does the build reliably fail if you keep the cache?

**Comment by XhmikosR** [comment]
Yeah it would seem so.

**Comment by XhmikosR** [comment]
I noticed the same thing with bootstrap too BTW.

**Comment by BanzaiMan** [comment]
What bootstrap?

**Comment by XhmikosR** [comment]
The [original]9https://github.com/travis-ci/travis-ci/issues/5092#issue-116461249) post is for Bootstrap but I notice the same failures in my other projects too.

**Comment by cvrebert** [comment]
Judging by https://travis-ci.org/twbs/bootstrap/builds/93292414 , it seems that the problem occurs even if the caching is only of `node_modules` and not RubyGems.

**Comment by RX14** [comment]
Same error: https://travis-ci.com/RX14/TRLauncher-v2/builds/19471269

**Comment by RX14** [comment]
Actually, this occurs whether caching is on or not.

**Comment by BanzaiMan** [comment]
Hello, so I looked at various reports of this kind in various places. It is quite likely that the `curl` warning message is false, but I am not sure how it can happen.

In all of the cases (except for one) I have seen, RVM finds a usable archive somewhere, and proceeds to install. But the `rvm` command exits from the `curl`-not-found error before the installation can finish, and the build fails on Travis.

**Comment by dmnd** [comment]
I get this error on my project only when I have the following:

```yaml
language: ruby
cache:
  bundler: true
  directories:
    - $HOME/.nvm-flexport
    - node_modules
rvm:
  - 2.2.3
```

If I stop caching node modules, rvm finds the binary version of Ruby. It's pretty frustrating because I'm trying to cache `node_modules` as `npm install` takes a very long time. And caching `node_modules` really does speed up `npm install`! Unfortunately due to this strange rvm/curl issue I end up losing all the time I gained by having to build Ruby :snail: :cry:

**Comment by cvrebert** [comment]
@BanzaiMan Would it be possible for you to dive into a Travis VM of a broken build and step through the RVM shell script to see what's wrong? The fact that the bug somehow involves Casher makes it harder for us regular users to debug ourselves.

**Comment by BanzaiMan** [comment]
@cvrebert Unfortunately not. The VMs (and containers) are short lived and currently there is no easy way to connect to a live build.

**Comment by cvrebert** [comment]
Perhaps it's a Casher bug? (https://github.com/travis-ci/casher/pull/19 )
I'm grasping at straws.

**Comment by dmnd** [comment]
I successfully worked around this bug like this:

```yaml
cache:
  directories:
    - node_modules_cache_hack

# snip

before_script:
  - mv node_modules_cache_hack node_modules
  - npm install --production

# snip


script:
  - npm test
  - mv node_modules node_modules_cache_hack
```

I guess there's still a chance that the `node_modules` directory is unrelated, and I'm just getting luck, but I'm pretty sure this is a workaround. Others here: please try it yourself!

**Comment by defsprite** [comment]
We're seeing this as well - yet only on projects that have `node_modules`. Only clue I've found:

- `node_modules/.bin` is on `$PATH`, prior to standard bin directories 
- in our case there is `node_modules/.bin/which`

As `rvm` uses `which` to detect `curl` might this be the issue? (see above and https://github.com/rvm/rvm/blob/master/scripts/functions/support#L262)

Not sure how to verify this though. It should probably work even with that version of `which`.

**Comment by dmnd** [comment]
:+1: nice sleuthing @defsprite, I have `node_modules/.bin/which` too.

**Comment by dmnd** [comment]
you could verify by getting rid of your dependency on the `which` module then seeing if rvm works.

**Comment by cvrebert** [comment]
Unfortunately, `grunt` depends on `which`, so that's rather difficult.

**Comment by defsprite** [comment]
`node_modules` is not checked in, and `which` is depended on a lot, according to https://libraries.io/npm/which

**Comment by cvrebert** [comment]
The builds are indeed using node-which instead of the system's which, although node-which is able to locate curl just fine. Excerpt from [build log](https://travis-ci.org/twbs/bootstrap/jobs/102452538):
```
$ which which || echo WHICH_LOST_ITSELF
/home/travis/build/twbs/bootstrap/node_modules/.bin/which
$ ./node_modules/.bin/which which || echo NODE_WHICH_LOST_ITSELF
/home/travis/build/twbs/bootstrap/node_modules/.bin/which
$ ./node_modules/.bin/which curl || echo NODE_NO_CURL
/usr/bin/curl
```

**Comment by defsprite** [comment]
Even weirder:

```
$ __rvm_which curl
/usr/bin/curl
```

**Comment by cvrebert** [comment]
Removing `node_modules

**Comment by cvrebert** [comment]
Removing `./node_modules/.bin` from the `$PATH` [indeed seems to avoid the problem](https://travis-ci.org/twbs/bootstrap/jobs/102460962).

**Comment by defsprite** [comment]
What puzzles me is that  – as far as I understand `rvm` – downloading from source also uses `curl` via `fetch_command`. So it shows up eventually when using `./node_modules/.bin/which curl`.

Fascinating.

**Comment by defsprite** [comment]
In my case I cannot modify path before install happens, but it helped to add the following to `.travis.yml`:

```
before_cache:
  - rm -rf ./node_modules/.bin
```

Of course, it only works if you're not relying the directory being there, which we don't.

**Comment by boxi13** [comment]
Col mu, I sipk elidel Englis, super italian lenguch.
12. jan. 2016 01:53 je oseba "Desmond Brand" <notifications@github.com>
napisala:

**Comment by vidmantas** [comment]
We're experiencing the same issue, but none of the mentioned workaround works, because we're using Ruby build and multiple workers.

**Comment by Fire-Dragon-DoL** [comment]
I'm currently using

```bash
before_install:
  # Dirty hack for https://github.com/travis-ci/travis-ci/issues/5092
  - >
    PATH=`echo $PATH | sed "s/\.\/node_modules\/\.bin//g"`;
    export PATH=$PATH:./node_modules/.bin;
```

As a dirty hack to avoid the issue. Notice that I found **two** node_modules path in `PATH`)

**Comment by vvo** [comment]
I did not change a thing on my travis.yml but all my builds on https://github.com/algolia/places/blob/master/.travis.yml.

I always get "RVM requires 'curl'. Install 'curl' first and try again."

**Comment by vvo** [comment]
Why is ./node_modules/.bin even on the PATH? This means any dependency installed without -g in npm will end up being global. This seems like a bad shortcut no?

**Comment by Fire-Dragon-DoL** [comment]
@vvo no, not necessarily, actually on a per-project basis it's great to have local executables. That being said, it should be an opt-in behavior, not the opposite

**Comment by vvo** [comment]
> actually on a per-project basis it's great to have local executables

With npm, if you need a global package, you do npm install -g.

If you want local dependencies, npm install dep.

If that dependency has an executable, as soon as you use npm scripts then if the script uses one of binaries, you do not have to provide ./node_modules/.bin/which, just "which" will works.

See https://docs.npmjs.com/misc/scripts#path

So basically this "feature" as to me was a shortcut that is now triggering issues.

**Comment by Fire-Dragon-DoL** [comment]
That still assumes you want to use npm scripts for everything. There are tools to allow you (in general) to have executables in path for specific projects, (I'm looking at `direnv`), so it's not like this was invented here or anything. I find much more reasonable to have executable related to my project appear on my path when I'm working on that project, and other people too do that since tools for that purpose exist.

I don't need any global packages beside tools that I use system wide, so it would be really disappointing having to maintain a separate list of global packages that needs to be installed.

**Comment by vvo** [comment]
Just pointing out that the default travis behavior is to considerer local project dependencies as being globals while it's not the case when you npm install on your machine. Making issues like this one more frequent than usually and hard to debug because of the unexpected behavior.

**Comment by Fire-Dragon-DoL** [comment]
@vvo oh as I said I definitely agree with you on that, it should be an opt-in feature, I have no idea why it's  added to PATH by default, not to mention added in wrong order.

**Comment by p3drosola** [comment]
This solves it for me

```yml
before_install:
  # which causes a problem with rvm curl detection, so we're moving it out of the way for a minute
  - mv node_modules/.bin/which node_modules/.bin/which.backup
  - rvm install ruby-2.3.0 --binary && rvm use ruby-2.3.0
  - mv node_modules/.bin/which.backup node_modules/.bin/which

**Comment by Kjir** [comment]
I confirm the issue and the simplest workaround is the latest one posted by @p3drosola

**Comment by pkuczynski** [comment]
Has this issue been already solved or RVM needs to prepare solution?

**Comment by BanzaiMan** [comment]
This is an issue with the use of nodejs-which. Unless your dependency tree removes it, the issue will persist, even if you don't use RVM.

**Comment by pkuczynski** [comment]
What if RVM would modify PATH removing node_modules/.bin and restore afterwards?

**Comment by pkuczynski** [comment]
Reported this issue to node-which (https://github.com/npm/node-which/issues/41). Lets see if they will be able to come up with some solution...

**Comment by Fire-Dragon-DoL** [comment]
I'm confused. Travis is injection stuff into PATH and breaking builds in this way, why node-which should fix anything? Shouldn't this be addressed from travis, stop injecting the path and developers should do that IF they want to?
I mean, node-which can definitely change the executable name, but still Travis is overstepping by injecting node_modules/.bin without even an option

**Comment by pkuczynski** [comment]
Is it Travis doing this or rather node?

**Comment by Fire-Dragon-DoL** [comment]
Node is not adding anything to PATH, sure it install packages, but it's travis the .bin directory to path

**Comment by BanzaiMan** [comment]
Are you suggesting that we should _not_ add `node_modules/.bin` to `$PATH`?

**Comment by pkuczynski** [comment]
This I have no idea, but I also tend to think that it should not be added, as this is not the default node behaviour (as I found out).

**Comment by Fire-Dragon-DoL** [comment]
Yes I'm suggesting that it shouldn't be added. That behavior is only present in travis, in any other environment is not present. People can either add `.bin` to PATH or use `"$(npm bin)/which"` to use the executables in there. It's actually what is commonly done.

**Comment by pkuczynski** [comment]
I agree with @Fire-Dragon-DoL. @BanzaiMan do you think we can get it fixed soon?

**Comment by isaacs** [comment]
> Yes I'm suggesting that it shouldn't be added. That behavior is only present in travis, in any other environment is not present. People can either add .bin to PATH or use "$(npm bin)/which" to use the executables in there. It's actually what is commonly done.

That is slightly incorrect, @Fire-Dragon-DoL.

Whenever you use an npm script (including `npm test` and friends), npm will put `node_modules/.bin` in the `PATH` environment for the subshell.  Without this, consider how `"scripts": { "test": "tap test/*.js" }` would work while also guaranteeing that when `test/foo.js` does `require("tap")` it gets the same version that is running its tests.  (Especially in a world where conflicting dependencies are resolved by using nested folders.)

Of course, I'm aware that there are many clever approaches that are different from what exist.  I've implemented several of them.  This is likely not the place to discuss them.  The barrier to entry is low, if you feel like creating your own npm client, and I encourage you to do this, if only to gain empathy for those who maintain such tools.

As for why `node-which` isn't finding `curl`, that's a really curious issue!

Can someone encountering this issue head over to https://github.com/npm/node-which/issues/41 and let me know what your `$PATH` environment variable is and what `ls -laF $(which curl)` reports?  Also, is `which curl` outputting anything?  Exiting with an non-zero error status code?  Anything you can share would be very helpful, since I can't reproduce this on my own system.

Thanks!

**Comment by Fire-Dragon-DoL** [comment]
@isaacs that's actually very interesting to know, wasn't aware of that `npm` behavior, I always append `$(npm bin)` to any command in the scripts!

However, isn't adding stuff to PATH still be a decision the developer should make, instead of travis? (beside the npm behavior, which is the same everywhere)

**Comment by pkuczynski** [comment]
@isaacs can you post the same request for more details in the https://github.com/npm/node-which/issues/41? I will try to get some people to respond to you then...

**Comment by isaacs** [comment]
https://github.com/npm/node-which/issues/41#issuecomment-263427113

Is it at all possible that this is either a race condition in Travis's setup, or an RVM bug?  It seems to be happening outside of npm scripts, where `node-which` wouldn't be a factor.

**Comment by pkuczynski** [comment]
Seems that node-which does not see a way of fixing this issue (https://github.com/npm/node-which/issues/41), so the only think which can be done is change path on travis to exclude node-which...

**Comment by kvz** [comment]
> Is it at all possible that this is either a race condition in Travis's setup, or an RVM bug? It seems to be happening outside of npm scripts, where node-which wouldn't be a factor.

The problem is where people cache `node_modules`, meaning upon next run, npm's `which` is available before anything else, and RVM happily uses that to establish if `curl` is present - vs the the system `which`.

**Comment by pkuczynski** [comment]
This should be fixed by Travis, as we have no way of fixing this on RVM side. Unless someone have an idea how....?

**Comment by kvz** [comment]
I don't think it's very common that people add `node_modules/.bin` to their `PATH`. I think it's more common that people add dependencies to their npm scripts, and then run those scripts. e.g.:

```json
...
"scripts": {
  "test": "mocha"
}
...
```

```bash
npm run test
```

I don't think people expect `mocha` to be present in their shell unless they installed it globally via `npm install -g mocha`.

It might be BC breaking for the few that relied on it, but (easily said ofc) to me that's probably worth it to make this right, and avoid more breakage for others.

**Comment by XhmikosR** [comment]
For what is worth, I too agree with @kvz.

Travis shouldn't do this in the first place. And yes, some breakage might occur, but it's all for the best outcome IMO.

**Comment by pkuczynski** [comment]
Travis guys, can you remove adding node_modules to the path?

**Comment by BanzaiMan** [comment]
Hello. Our attempt to remove `./node_modules/.bin` from `$PATH` negatively affected many existing builds.

This issue is going to cause pain in non-negligible number of users either way, unfortunately.

**Comment by Fire-Dragon-DoL** [comment]
@BanzaiMan It broke many existing builds (based on this bug too) on the other side too, so at least assume the correct default and let people fix it

**Comment by BanzaiMan** [comment]
See also: #4862.

**Comment by vvo** [comment]
This issue is really long, the tl;dr; is:
- travis by default puts ./node_modules/.bin as globally available binaries (modifying PATH)
- some projects might rely on that (like Ember)
- there's no decision for now taken
- there's an open PR https://github.com/travis-ci/travis-build/pull/903 and https://github.com/travis-ci/travis-cookbooks/pull/786

There's a way to solve it for you:

```yml
before_cache:
  - rm -f ./node_modules/.bin/which # <-- workaround for https://github.com/travis-ci/travis-
```

**Comment by szimek** [comment]
Wouldn't it be possible to add a setting like:
```
i_do_not_want_all_my_node_modules_binaries_to_be_global_thank_you: true
```
(or something similar 😉) that would cause `./node_modules/.bin` to be _not_ added to `PATH`? This could become the default behavior e.g. when a new image is released in the future.

**Comment by ankon** [comment]
> [An option in `.travis.yml`] that would cause ./node_modules/.bin to be not added to PATH? This could become the default behavior e.g. when a new image is released in the future.

That would be much appreciated indeed!

**Comment by ankon** [comment]
This problem still exists, and is still as annoying as ever -- please keep it open :)

**Comment by ankon** [comment]
Bot: Can you leave this open and linked to #1263 please? :)

**Comment by crequena-babbel** [comment]
As workaround for this bug I'm doing this in the .travis.yml :

`PATH="/bin:$PATH" rvm use $(< .ruby-version) --install --binary --fuzzy`
